src/quacc/calculators/espresso/utils.py

```diff
@@ -0,0 +1,154 @@
+            if key in parameters.get(section, {}):
+                sec_list[key] = parameters[section].pop(key)
+            if key in parameters:
+                sec_list[key] = parameters.pop(key)
+            if key in kwargs:
+                sec_list[key] = kwargs.pop(key)
+            # Check if there is a key(i) version (no extra parsing)
+            for arg_key in list(parameters.get(section, {})):
+                if arg_key.split('(')[0].strip().lower() == key.lower():
+                    sec_list[arg_key] = parameters[section].pop(arg_key)
+            cp_parameters = parameters.copy()
+            for arg_key in cp_parameters:
+                if arg_key.split('(')[0].strip().lower() == key.lower():
+                    sec_list[arg_key] = parameters.pop(arg_key)
+            cp_kwargs = kwargs.copy()
+            for arg_key in cp_kwargs:
+                if arg_key.split('(')[0].strip().lower() == key.lower():
+                    sec_list[arg_key] = kwargs.pop(arg_key)
+
+        # Add to output
+        input_namelist[section] = sec_list
+
+    unused_keys = list(kwargs)
+    # pass anything else already in a section
+    for key, value in parameters.items():
+        if key in keys and isinstance(value, dict):
+            input_namelist[key].update(value)
+        elif isinstance(value, dict):
+            unused_keys.extend(list(value))
+        else:
+            unused_keys.append(key)
+
+    if warn and unused_keys:
+        warnings.warn('Unused keys: {}'.format(', '.join(unused_keys)))
+
+    return input_namelist
+
+
+def parse_pp_and_cutoff(config, atoms):
+    # if atoms are here, we compute the highest ecutwfc and ecutrho
+
+    # pseudopotentials should be the only special case
+    # in the sense that it contains additional information
+    if 'pseudopotentials' in config:
+        pp_dict = config['pseudopotentials']
+        unique_elements = list(set(atoms.symbols))
+        wfc_cutoff, rho_cutoff = 0, 0
+        pseudopotentials = {}
+        for element in unique_elements:
+            if pp_dict[element]['cutoff_wfc'] > wfc_cutoff:
+                wfc_cutoff = pp_dict[element]['cutoff_wfc']
+            if pp_dict[element]['cutoff_rho'] > rho_cutoff:
+                rho_cutoff = pp_dict[element]['cutoff_rho']
+            pseudopotentials[element] = \
+                pp_dict[element]['filename']
+    else:
+        return {}
+    tmp_input_data = {
+        'system': {
+            'ecutwfc': wfc_cutoff,
+            'ecutrho': rho_cutoff}}
+    return {
+        'pseudopotentials': pseudopotentials,
+        'input_data': tmp_input_data}
+
+
+def namelist_to_string(parameters):
+    # This function aim to write basic input keywords to a file
+    # For additional cards (binary specifics) this is down in the child class
+    # Stolen from ase.io.espresso,
+    # maybe this could be made to a function "write_namelist"
+    # and imported here? See with the ASE gods.
```

Comment:
I am an ASE maintainer. I can review and approve an MR for this.

Reply:
I created a MR on ase master, also to add the 'keys' parameter the construct_namelist function, so that we avoid copying code here. This should not affect ASE at all.

Reply:
Great, thanks! I will review that and get it merged. In the meantime, you can keep what you have here and when it's merged in ASE we can swap it out.